pallets/asset/src/identifier.rs

```diff
@@ -0,0 +1,176 @@
+use codec::{Decode, Encode};
+use core::convert::TryInto;
+use sp_std::vec::Vec;
+
+#[derive(Encode, Decode, Clone, Debug, PartialEq, Eq)]
+pub enum Identifier {
+    CUSIP([u8; 9]),
+    CINS([u8; 9]),
+    ISIN([u8; 12]),
+    LEI([u8; 20]),
+    EMPTY,
+}
+
+impl Default for Identifier {
+    fn default() -> Self {
+        Identifier::EMPTY
+    }
+}
+
+impl Identifier {
+    pub fn cusip(bytes: [u8; 9]) -> Option<Identifier> {
+        (cusip_checksum(&bytes[..8]) == bytes[8] - b'0').then_some(Identifier::CUSIP(bytes))
+    }
+
+    pub fn cins(bytes: [u8; 9]) -> Option<Identifier> {
+        (cusip_checksum(&bytes[..8]) == bytes[8] - b'0').then_some(Identifier::CINS(bytes))
+    }
+
+    pub fn isin(bytes: [u8; 12]) -> Option<Identifier> {
+        let (s1, s2) = bytes
+            .iter()
+            .map(|b| byte_value(*b))
+            .flat_map(|b| if b > 9 { Vec::from([b / 10, b % 10]) } else { Vec::from([b]) })
+            .rev()
+            .enumerate()
+            .fold((0, 0), |(mut s1, mut s2), (i, digit)| {
+                if i % 2 == 0 {
+                    s1 += digit;
+                } else {
+                    s2 += 2 * digit;
+                    if digit >= 5 {
+                        s2 -= 9;
+                    }
+                }
+                (s1, s2)
+            });
+        ((s1 + s2) % 10 == 0).then_some(Identifier::ISIN(bytes))
+    }
+
+    pub fn lei(bytes: [u8; 20]) -> Option<Identifier> {
+        (lei_checksum(bytes[..18].try_into().ok()?) == (bytes[18] - b'0') * 10 + (bytes[19] - b'0'))
+            .then_some(Identifier::LEI(bytes))
+    }
+}
+
+fn cusip_checksum(bytes: &[u8]) -> u8 {
+    let total: usize = bytes
+        .iter()
+        .copied()
+        .map(byte_value)
+        .enumerate()
+        .map(|(i, v)| v << (i % 2))
+        .map(|v| (v / 10) + v % 10)
+        .map(|i| i as usize)
+        .sum();
+    ((10 - (total % 10)) % 10) as u8
+}
+
+fn lei_checksum(bytes: [u8; 18]) -> u8 {
+    let (total, _) = bytes
+        .iter()
+        .rev()
+        .fold((0u128, 0usize), |(mut total, mut i), b| {
+            total += byte_value(*b) as u128 * 10u128.pow(i as u32);
```

Comment:
Compute `byte_value`  by using `.map(...)` so that you don't have to compute it twice.